File: backend/app/schemas/report.py
"""Pydantic schemas for bug/feature reports."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ReportType = Literal["bug", "feature"]
ReportStatus = Literal["open", "in_progress", "closed"]
ReportSeverity = Literal["info", "warning", "error"]
ReportSource = Literal["frontend", "backend"]


class ReportCreate(BaseModel):
    """User submission of a bug report or feature request."""

    type: ReportType
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1, max_length=10000)
    metadata: dict[str, Any] | None = None
    conversation_id: str | None = Field(default=None, max_length=36)
    severity: ReportSeverity | None = None
    source: ReportSource | None = None


class ReportStatusUpdate(BaseModel):
    """Admin triage: move a report through its status flow."""

    status: ReportStatus


class ReportOut(BaseModel):
    """Report as returned by the API."""

    id: str
    user_id: str
    type: ReportType
    title: str
    description: str
    metadata: dict[str, Any] | None = Field(default=None, validation_alias="metadata_")
    conversation_id: str | None
    severity: ReportSeverity | None
    source: ReportSource | None
    status: ReportStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: lib/features/reports/models/report.dart
/// A user-submitted bug report or feature request (idea 14).
///
/// [type] is `bug` or `feature`; [status] flows `open` → `in_progress` →
/// `closed` and is admin-controlled.
class Report {
  const Report({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.description,
    this.metadata,
    this.conversationId,
    this.severity,
    this.source,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String userId;
  final String type;
  final String title;
  final String description;
  final Map<String, dynamic>? metadata;
  final String? conversationId;
  final String? severity;
  final String? source;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Report.fromJson(Map<String, dynamic> json) => Report(
    id: json['id'] as String,
    userId: json['user_id'] as String,
    type: json['type'] as String,
    title: json['title'] as String,
    description: json['description'] as String,
    metadata: (json['metadata'] as Map?)?.cast<String, dynamic>(),
    conversationId: json['conversation_id'] as String?,
    severity: json['severity'] as String?,
    source: json['source'] as String?,
    status: json['status'] as String,
    createdAt: DateTime.parse(json['created_at'] as String),
    updatedAt: DateTime.parse(json['updated_at'] as String),
  );
}
